File: backend/app/models/practice.py
from pydantic import BaseModel
from typing import Optional, List


class ProsodyMetrics(BaseModel):
    speech_rate: float = 0.0
    articulation_rate: float = 0.0
    pause_count: int = 0
    mean_pause_duration: float = 0.0
    long_pause_count: int = 0
    f0_mean: float = 0.0
    f0_std: float = 0.0
    intonation_index: float = 0.0
    total_speech_time: float = 0.0
    total_duration: float = 0.0


class PronunciationIssue(BaseModel):
    word: str
    expected: Optional[str] = None
    timestamp: float = 0.0
    confidence: float = 0.0


class GrammarError(BaseModel):
    sentence: str
    correction: str
    explanation: str


class AnalysisScores(BaseModel):
    overall: float = 0.0
    grammar: float = 0.0
    fluency: float = 0.0
    pronunciation: float = 0.0
    prosody: float = 0.0


class AnalysisResult(BaseModel):
    transcript: str = ""
    word_timestamps: list = []
    prosody: ProsodyMetrics = ProsodyMetrics()
    pronunciation_issues: List[PronunciationIssue] = []
    is_read_aloud: Optional[bool] = None
    scores: AnalysisScores = AnalysisScores()
    errors: List[GrammarError] = []
    suggestions: List[str] = []
    llm_feedback: str = ""


class PracticeSessionResponse(BaseModel):
    id: int
    student_id: int
    exercise_id: Optional[int] = None
    transcript: Optional[str] = None
    duration_seconds: Optional[float] = None
    created_at: Optional[str] = None
    result: Optional[dict] = None
    exercise_title: Optional[str] = None

File: frontend/src/types/practice.ts
export interface ProsodyMetrics {
  speech_rate: number
  articulation_rate: number
  pause_count: number
  mean_pause_duration: number
  long_pause_count: number
  f0_mean: number
  f0_std: number
  intonation_index: number
  total_speech_time: number
  total_duration: number
}

export interface PronunciationIssue {
  word: string
  expected?: string | null
  timestamp: number
  confidence: number
}

export interface GrammarError {
  sentence: string
  correction: string
  explanation: string
}

export interface AnalysisScores {
  overall: number
  grammar: number
  fluency: number
  pronunciation: number
  prosody: number
}

export interface AnalysisResult {
  transcript: string
  word_timestamps: Array<{ word: string; start: number; end: number }>
  prosody: ProsodyMetrics
  pronunciation_issues: PronunciationIssue[]
  is_read_aloud: boolean | null
  scores: AnalysisScores
  errors: GrammarError[]
  suggestions: string[]
  strengths: string[]
  areas_to_improve: string[]
  llm_feedback: string
  prosody_feedback: string
  pronunciation_feedback: string
}

export interface PracticeSession {
  id: number
  student_id: number
  exercise_id?: number | null
  transcript?: string | null
  duration_seconds?: number | null
  created_at?: string | null
  result?: Record<string, unknown> | null
  exercise_title?: string | null
}
